### DISCUSSION TITLE
Children components events

### DISCUSSION THREAD

**Discussion by lamtech58** [arkivanov/Decompose#585] [Q&A] [answered]
I just started using this lib , after only 5 days I found myself handling a lots of children components events from the root component, can you please guide me on how to do that better way?

Sorry for my poor English, hope you understand what I mean.

**Comment by arkivanov** [comment]
Maybe you could provide some code examples?

**Reply by lamtech58** [reply]
sure, here is a google sign in component exemple :

```kotlin
override val googleSignInComponent = GoogleSignInComponent(
        childContext(key = "google sinIn comp"),
        onEvent = { event ->
            when (event) {
                GoogleSignInComponentEvent.OnCollapseGoogleSignIn -> {
                    setBottomSheetType(RootComponent.BottomSheetType.Idle)
                }

                is GoogleSignInComponentEvent.OnSuccessfullySignIn -> {
                    setBottomSheetType(RootComponent.BottomSheetType.Idle)

                    rootEvent(
                        RootComponentEvent.OnShowSnackBar(event.state)
                    )
                }

                is GoogleSignInComponentEvent.OnShowSnackBar -> {
                    rootEvent(
                        RootComponentEvent.OnShowSnackBar(event.state)
                    )
                }
            }
        }
    )
```

in every component child i have to handle its events with  RootComponentEvent and i think it could be better

**Reply by arkivanov** [reply]
It looks like your GoogleSignInComponent is displayed in a bottom sheet. Perhaps, you could leverage Child Slot navigation model, which should reduce some boilerplate.

For snackbar, you could create a common interface like `SnackbarController` with a function like `show`. And then just pass that to all child components that need to display snackbars.

Also, I think it would look better if you use just normal callbacks instead of sealed class events. E.g. that code could be:

```kotlin
override val googleSignInComponent = GoogleSignInComponent(
        childContext(key = "google sinIn comp"),
        onCollapseGoogleSignIn = {},
        onSuccessfullySignIn = {},
        onShowSnackBar = {},
)
```

So you don't need sealed classes and `when` switch.

**Reply by lamtech58** [reply, marked as answer]
Thanks for your time, I'll do as you suggested
